src/Types/OnChainData.cs

```diff
@@ -0,0 +1,144 @@
+    /// </summary>
+    /// <param name="data">
+    /// Text to store on-chain represented as a <see cref="string"/> whose CBOR encoding will be used for the data to registered on-chain.
+    /// </param>
+    /// <param name="conformanceMode">
+    /// The conformance mode to use for encoding and decoding CBOR data.
+    /// </param>
+    /// <exception cref="ArgumentException">The resulting CBOR encoded string exceeds <see cref="MaxLength"/> bytes.</exception>
+    /// <exception cref="ArgumentException">The supplied string is not a valid UTF-8 encoding and this is not permitted under the current conformance mode.</exception>
+    /// <exception cref="InvalidOperationException">The written CBOR data is not accepted under the current conformance mode.</exception>
+    public static OnChainData FromTextEncodeAsCBOR(
+        string data,
+        CborConformanceMode conformanceMode = CborConformanceMode.Strict
+    )
+    {
+        var encoder = new CborWriter(conformanceMode);
+        encoder.WriteTextString(data);
+        var encodedBytes = encoder.Encode();
+        return From(encodedBytes);
+    }
+
+    /// <summary>
+    /// Creates an instance from byte array.
+    /// </summary>
+    /// <param name="data">The data to be registered on-chain represented as a byte array.</param>
+    /// <exception cref="ArgumentException">The length of the supplied data exceeds <see cref="MaxLength"/>.</exception>
+    public static OnChainData From(byte[] dataAsBytes)
+    {
+        if (dataAsBytes.Length > MaxLength)
+        {
+            throw new ArgumentException(
+                $"Size of a data is not allowed to exceed {MaxLength} bytes."
+            );
+        }
+
+        return new OnChainData((byte[])dataAsBytes.Clone());
+    }
+
+    /// <summary>
+    /// Try to decode the data to be registered on-chain as a single CBOR encoded string.
+    /// </summary>
+    /// <returns>
+    /// A <c>string</c> corresponding to the decoded data if it contained
+    /// a single CBOR encoded string, and <c>null</c> otherwise.
+    /// </returns>
+    public string? TryCborDecodeToString()
+    {
+        var encoder = new CborReader(this._value);
+        try
+        {
+            var textRead = encoder.ReadTextString();
+            if (encoder.BytesRemaining == 0)
+            {
+                return textRead;
+            }
+        }
+        catch (CborContentException) { }
+        catch (InvalidOperationException) { }
+        return null;
+    }
+
+    /// <summary>
+    /// Get the on-chain data in the binary format expected by the node.
+    ///
+    /// That is, represented as a byte array with the length of the array
+    /// prepended as a 16-bit unsigned integer in big-endian format.
+    /// </summary>
+    public byte[] GetBytes()
+    {
+        using var memoryStream = new MemoryStream(sizeof(ushort) + this._value.Length);
+        memoryStream.Write(Serialization.GetBytes((ushort)this._value.Length));
```

Comment:
```suggestion
        using var memoryStream = new MemoryStream(sizeof(ushort) + this._value.Length);
        memoryStream.Write(Serialization.GetBytes(sizeof(ushort)));
```

wouldn’t this be more consistent?

Reply:
No, that would write the constant `sizeof(ushort)` to the stream. We want to write the length of the data byte array as a `ushort`.